src/Memory/include/Memory/MemoryInfoProducer.h

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_MEMORY_INFO_PRODUCER_H_
+#define ORBIT_MEMORY_INFO_PRODUCER_H_
+
+#include <atomic>
+#include <thread>
+
+#include "Memory/MemoryInfoListener.h"
+#include "capture.pb.h"
+
+namespace orbit_memory {
+class MemoryInfoProducer {
+ public:
+  explicit MemoryInfoProducer(uint64_t memory_sampling_period_ns)
+      : sampling_period_ns_(memory_sampling_period_ns) {}
+
+  MemoryInfoProducer(const MemoryInfoProducer&) = delete;
+  MemoryInfoProducer& operator=(const MemoryInfoProducer&) = delete;
+  MemoryInfoProducer(MemoryInfoProducer&&) = delete;
+  MemoryInfoProducer& operator=(MemoryInfoProducer&&) = delete;
+
+  ~MemoryInfoProducer() { Stop(); }
+
+  void SetListener(MemoryInfoListener* listener) { listener_ = listener; }
+
+  void Start();
+  void Stop();
+
+ private:
+  void Run();
+  void ProduceTotalMemoryUsageInfoAndSendToListener();
+
+  uint64_t sampling_period_ns_;
+  MemoryInfoListener* listener_ = nullptr;
+  std::shared_ptr<std::atomic<bool>> exit_requested_ = std::make_unique<std::atomic<bool>>(true);
```

Comment:
Do you need this to be wrapped inside a smart pointer?

Reply:
Make sense. I just removed the smart pointer. Thanks for pointing out this. Please see the commit *[fixup] Not use smart pointer in MemoryInfoProducer*.